src/rules/noConditionalAssignmentRule.ts

```diff
@@ -56,39 +56,48 @@ function walk(ctx: Lint.WalkContext<void>) {
-                return check((node as ts.IfStatement).expression),
-                       cb((node as ts.IfStatement).thenStatement),
-                       maybeCallback(cb, (node as ts.IfStatement).elseStatement);
+                check((node as ts.IfStatement).expression);
+                cb((node as ts.IfStatement).thenStatement);
+                maybeCallback(cb, (node as ts.IfStatement).elseStatement);
+                return;
             case ts.SyntaxKind.DoStatement:
             case ts.SyntaxKind.WhileStatement:
-                return check((node as ts.DoStatement | ts.WhileStatement).expression),
-                       cb((node as ts.IterationStatement).statement);
+                check((node as ts.DoStatement | ts.WhileStatement).expression);
+                cb((node as ts.IterationStatement).statement);
+                return;
             case ts.SyntaxKind.ForStatement:
-                return maybeCallback(cb, (node as ts.ForStatement).initializer),
-                       maybeCallback(check, (node as ts.ForStatement).condition),
-                       maybeCallback(cb, (node as ts.ForStatement).incrementor),
-                       cb((node as ts.ForStatement).statement);
+                maybeCallback(cb, (node as ts.ForStatement).initializer);
+                maybeCallback(check, (node as ts.ForStatement).condition);
+                maybeCallback(cb, (node as ts.ForStatement).incrementor);
+                cb((node as ts.ForStatement).statement);
+                return;
         }
         if (checking !== 0) {
             switch (kind) {
                 case ts.SyntaxKind.BinaryExpression:
                     if (isAssignmentKind((node as ts.BinaryExpression).operatorToken.kind)) {
                         ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
                     }
-                    return cb((node as ts.BinaryExpression).left), cb((node as ts.BinaryExpression).right);
+                    cb((node as ts.BinaryExpression).left);
+                    cb((node as ts.BinaryExpression).right);
+                    return;
                 case ts.SyntaxKind.ParenthesizedExpression:
                 case ts.SyntaxKind.NonNullExpression:
                 case ts.SyntaxKind.AsExpression:
                 case ts.SyntaxKind.TypeAssertionExpression:
-                    return cb((node as ts.AssertionExpression | ts.NonNullExpression | ts.ParenthesizedExpression).expression);
+                    cb((node as ts.AssertionExpression | ts.NonNullExpression | ts.ParenthesizedExpression).expression);
+                    return;
                 case ts.SyntaxKind.PrefixUnaryExpression:
-                    return cb((node as ts.PrefixUnaryExpression).operand);
+                    cb((node as ts.PrefixUnaryExpression).operand);
+                    return;
                 default:
-                    return noCheck(node);
+                    noCheck(node);
+                    return;
```

Comment:
it's not necessary to change the 3 returns above.

Reply:
That's true. Changed back.

I changed them for consistency with the others; I don't like that `return foo()` implies that `foo()` actually returns a value, rather than has only side effects.

Reply:
I introduced that pattern back then thinking this might benefit from tail call optimization. I haven't measured if it actually makes a difference though.